### ISSUE TITLE
ZIP-221: Validate chain history commitments in the finalized state

### ISSUE THREAD

**Issue by mpguerra** [ZcashFoundation/zebra#2134]
## Is your feature request related to a problem?

Once we have integrated the required structures for ZIP-221 from librustzcash (#2132), we should use these to update our implementation of Zebra state for the finalized tip, and validate the chain history commitments for finalized blocks.

### Maintaining History

For pre-Canopy blocks:
- [ ] There is no history merkle mountain range (MMR)
  - history commitments were added to Zcash at the Heartwood network upgrade
  - Zebra uses checkpoints for Heartwood, rather than the history commitment

For the first block in each network upgrade:
- [ ] Delete the old network upgrade's history MMR cache
- [ ] Create a new history MMR when Zebra reaches a network upgrade activation block
- [ ] Add the network upgrade activation block to the history MMR

For later blocks:
- [ ] Get the previous block's history MMR from the state
  - add a new `ChainHistory` request to the state to get the history MMR data
- [ ] Add the new block to the previous block's history MMR
- [ ] Prune unused cached state

### Finalized state

Cache management:
- [ ] write a function which works out when cached nodes aren't needed any more
  - `Tree::root`, `IndexedNode::left`/`right`, `Tree::resolve_link` let us walk the tree
  - get the indexes of all stored and generated nodes that are still used
  - delete old `Stored` entries that aren't used any more
  - add new `Generated` entries
  - we might want to create a `CacheAction` enum with `Add` and `Delete` variants

Database format:
- [ ] Add a new RocksDB column family for the history MMR data
  - key `Height` and value `Entry::to_bytes` / `from_bytes`
  - calculate `height` using `index + network_upgrade.activation_height(network)`
  - this avoids bugs where we confuse cached state from different network upgrades
- [ ] Increment Zebra's state version

FinalizedBlock type:
- [ ] Add a history MMR data field to `FinalizedBlock`
- [ ] Create each `FinalizedBlock` using the calculated history MMR

For each FinalizedBlock:
- [ ] In `commit_finalized_direct`, validate that the history commitment from the history MMR matches the history commitment in the block header
  - This check is required from NU5 onwards, even during checkpointing, so we can bind the authorizing data to the block header
  - It's optional during checkpointing for Canopy, but if we do it, we'll get better test coverage for the history code
- [ ] If validation fails, return a new error. This error is network-upgrade specific:
  - Canopy: the block header binds all the transaction data, so this is probably a bug in Zebra. Panic.
  - NU5 and later: the authorizing data has been modified: checkpoint or full block validation succeeded, but the authorizing data doesn't match the authorizing data in the transaction that was mined into the block
- [ ] Add/Delete cached MMR state, as part of that block's **database** transaction

## Additional context

See https://github.com/zcashfoundation/zebra/issues/2091#issuecomment-832311598 for further details

**Comment by teor2345** [comment]
We need the sapling tree root for this change, so testing and full implementation are blocked by the sapling part of #1287.

**Comment by mpguerra** [comment]
Moving to Sprint 13 for now

**Comment by conradoplg** [comment]
Here's a few points which I'm not sure about the design.

# Storing the tree

To load a HistoryTree we need:
- `peaks`: A list of peak nodes (index, Entry)
  - Note that the index is the index of the node in the array representation of the tree.
    It is _not_ the height of the block from where the node was computed. There is
    a 1:1 correspondence only between blocks and _leaf_ nodes!
- `size`: The total number of nodes in the tree (often larger than the number of peaks)
- `current_height`: The height of the most recent block added to the tree.

This implies needing to change the design in `0005-state-updates.md`. Currently it lists:

```
| `history_incremental` | `BE32(height)`        | `zcash_history::Entry`         | Delete  |
```

> - The history tree indexes its peaks using blocks since the last network upgrade.
>   But we map those peak indexes to heights, to make testing and debugging easier.

That won't work since nodes are not indexed by height. I see two approaches:

## Store each node separatedly

Simply use the index instead of the height. This has the following implications:

- We would need a new column to store `size` and `current_height`. This column can be
  indexed by height and for each block we delete the previous and store the next, like
  we do for note commitment trees. How we would encode those values?
- `HistoryTree` would need to indicate when an upgrade happened (or we can simply
  recheck it in the state) in order to wipe the previous tree nodes.
  See "Upgrading the tree" below.
- We need to find out which nodes are no longer needed and which ones were added.
  See "Pruning the tree" below.

## Store the whole tree

Just like we currently do for note commitment trees. Note that the "whole tree" is just the list of peaks that has ~log(n) nodes. This has the following implications:

- Just store the whole data in a single column, like the note commitment trees
- More writes to the DB

# Pruning the tree

HistoryTree already prunes its nodes. But we need to redo that pruning in the state, unless we go for the "Store the whole tree" strategy. Possible options for that:

- We could change HistoryTree to return this information. The slight advantage
  is that it already has access to the nodes added when pushing a block;
  it would need only to list the pruned nodes.
- But we can also derive it from the `peaks` BTreeMap: build a set of old indices 
  and a set of new indices, and subtract them to find which old indices are not in the new
  set, and which new indices are not in the old set. Then, add/delete nodes
  accordingly.


# Upgrading the tree

As mentioned, HistoryTree can handle the network upgrades and the caller won't need to care (unless as mentioned above if it needs to delete the old nodes on upgrade).

This implies that we'll keep a `Option<HistoryTree>`, instantiate the tree on Heartwood activation, and keep working on it.

An alternative would be for the state to keep track of the network upgrades and recreate the HistoryTree when needed. Then we would be able to remove that logic from HistoryTree. However that logic would need to replicated in both finalized and non-finalized state.


# Validating the tree

I believe it would be better to move validation to a new ticket, just like we're doing for note commitment trees.

But I have some questions about the current design described in the issue:

>  Get the previous block's history MMR from the state
>    add a new ChainHistory request to the state to get the history MMR data

Can't we simply check it in the non-finalized state, when pushing a block into it?

>    In commit_finalized_direct, validate that the history commitment from the history MMR matches the history commitment in the block header

Likewise, shouldn't this be done in the non-finalized state? I assumed the finalized state only stores blocks that were already fully validated.

**Comment by teor2345** [comment]
> ## Store the whole tree
> 
> Just like we currently do for note commitment trees. Note that the "whole tree" is just the list of peaks that has ~log(n) nodes. This has the following implications:
> 
>     * Just store the whole data in a single column, like the note commitment trees
>     * More writes to the DB

Storing the whole tree seems like the simplest option for now.
We can open a ticket to optimise it later, if you'd like.

> # Validating the tree
> 
> I believe it would be better to move validation to a new ticket, just like we're doing for note commitment trees.

Sounds good!
Let's try to avoid doing too much in a single ticket or PR.

> But I have some questions about the current design described in the issue:
> 
> > Get the previous block's history MMR from the state
> > add a new ChainHistory request to the state to get the history MMR data
> 
> Can't we simply check it in the non-finalized state, when pushing a block into it?

The design in that ticket is pretty old now.

Since that design, we changed the non-finalized state to take a history tree.
So we can do history tree validation in the non-finalized state now.

> > In commit_finalized_direct, validate that the history commitment from the history MMR matches the history commitment in the block header
> 
> Likewise, shouldn't this be done in the non-finalized state? I assumed the finalized state only stores blocks that were already fully validated.

Unfortunately not, because we need to make sure that v5 transactions in checkpointed blocks haven't been modified.

`commit_finalized_direct` commits checkpointed blocks to the state.

The checkpoint verifier validates blocks by constructing a chain of block header hashes between checkpoints.
So to fully validate checkpointed blocks, we need to bind all the transaction data in the block to the block header.

For v1-v4 transactions, here's how we do that in the checkpoint verifier:
- the transaction id binds all the transaction data
- the merkle root in the transaction header binds the transaction ids

For v5 transactions, here's how we do that:

Checkpoint verifier:
- the transaction id binds the effects of the transaction
- the merkle root in the transaction header binds the transaction ids

Finalized state:
- the transaction authorization hash binds the signatures and proofs in the transaction 
- the block commitment binds the transaction authorization hashes and the chain history commitment

**Comment by conradoplg** [comment]
I think we can close this since validation has its own issue #2562
